sdk/storage/Azure.Storage.Files.Shares/src/PartitionedDownloader.cs

```diff
@@ -0,0 +1,392 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.IO;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Core.Pipeline;
+using Azure.Storage.Files.Shares.Models;
+
+namespace Azure.Storage.Files.Shares
+{
+    internal class PartitionedDownloader
+    {
+        /// <summary>
+        /// The client used to download the file.
+        /// </summary>
+        private readonly ShareFileClient _client;
+
+        /// <summary>
+        /// The maximum number of simultaneous workers.
+        /// </summary>
+        private readonly int _maxWorkerCount;
+
+        /// <summary>
+        /// The size of the first range requested (which can be larger than the
+        /// other ranges).
+        /// </summary>
+        private readonly long _initialRangeSize;
+
+        /// <summary>
+        /// The size of subsequent ranges.
+        /// </summary>
+        private readonly long _rangeSize;
+
+        public PartitionedDownloader(
+            ShareFileClient client,
+            StorageTransferOptions transferOptions = default)
+        {
+            _client = client;
+
+            // Set _maxWorkerCount
+            if (transferOptions.MaximumConcurrency.HasValue
+                && transferOptions.MaximumConcurrency > 0)
+            {
+                _maxWorkerCount = transferOptions.MaximumConcurrency.Value;
+            }
+            else
+            {
+                _maxWorkerCount = Constants.Blob.Block.DefaultConcurrentTransfersCount;
```

Comment:
we should do some benchmarks to know if the default value for shares is the same for blobs (probably applies to other defaults as well).